src/utils/coroutine-utils.ts

```diff
@@ -0,0 +1,87 @@
+import { coroutine } from "./coroutine";
+
+// TODO Write a better type for coroutine
+type Coroutine = () => IteratorResult<undefined, any>;
+// TODO: A better type for this
+type RollbackFunction = () => void;
+export type ClearFunction = () => void;
+type JobStartCallback = (
+  clearRollbacks: ClearFunction,
+  abortSignal: AbortSignal
+) => Generator<Promise<any> | CancelablePromise<any>, any, any>;
+export type Job = {
+  coroutine?: Coroutine;
+  fn: JobStartCallback;
+  abortController: AbortController;
+  rollbacks: RollbackFunction[];
+};
+
+export class JobRunner<T> {
+  jobs = new Map<T, Job>();
+  pendingStart: Job[] = [];
+
+  add(key: T, fn: JobStartCallback) {
+    if (this.jobs.has(key)) {
+      throw new Error(`Job already exists for key ${key}`);
+    }
+    const rollbacks: RollbackFunction[] = [];
+    const abortController = new AbortController();
+    const job = {
+      fn,
+      abortController,
+      rollbacks
+    };
+    this.pendingStart.push(job);
+    this.jobs.set(key, job);
+  }
+
+  has(key: T) {
+    return this.jobs.has(key);
+  }
+
+  stop(key: T) {
+    const job = this.jobs.get(key);
+    if (!job) return false;
+    job.abortController.abort();
+    for (let i = job.rollbacks.length - 1; i >= 0; i--) {
+      job.rollbacks[i]();
+    }
+    this.jobs.delete(key);
+    return true;
+  }
+
+  tick() {
+    this.pendingStart.forEach(job => {
+      const clearRollbacks = () => {
+        job.rollbacks.length = 0;
+      };
+      job.coroutine = coroutine(job.fn(clearRollbacks, job.abortController.signal), job.rollbacks);
+    });
+    this.pendingStart.length = 0;
+
+    this.jobs.forEach((job, eid) => {
+      if (job.coroutine!().done) {
+        this.jobs.delete(eid); // TODO Is this safe?
+      }
+    });
+  }
+}
+
+const $rollbackSymbol = Symbol("rollback");
+class CancelablePromise<T> {
+  promise: Promise<T>;
+  rollback: RollbackFunction;
+  $rollback = $rollbackSymbol;
+  constructor(promise: Promise<T>, rollback: RollbackFunction) {
+    this.promise = promise;
+    this.rollback = rollback;
+  }
+}
+
+export function isCancelablePromise<T>(c: any): c is CancelablePromise<T> {
+  return c.$rollback === $rollbackSymbol;
```

Comment:
It was late when we wrote this. I think what we wanted was
```ts
c[$rollbackSymbol] === true
```

with
```ts
static [$rollbackSymbol] = true;
```

What we have works, but the whole point of using a symbol for this sort of thing is that you shouldn't be able to get a reference to it externally. (this really doesn't matter)

Reply:
We tried this and it didn't work, so we changed it. I went back and tried some variations. This is what I've found to work:
```ts
  return c.__proto__.constructor[$rollbackSymbol] === true;
```
or (without `$rollbackSymbol` at all) 
```ts
c.__proto__ === CancelablePromise.prototype
```

Is either of these better? For the first, I probably also need to guard against `__proto__` or `constructor` being `undefined` since `c` is `any`.